### PR TITLE
fix(migrations): Add in a feature to force a lock timeout to be set on all migration operations

### PR DESCRIPTION
**Pull request by wedamija**
We had an incident where a migration couldn't acquire a lock to modify `sentry_groupedmessage` and took down production. This shouldn't have happened, since we rely on django-zero-downtime-migrations to set a lock acquisition timeout on all operations.

Unfortunately, it doesn't set it on `RunSQL` or `RunPython` ops. There are other operations that it considers safe that it also doesn't set it on, but those aren't an issue.

This pr adds `ZERO_DOWNTIME_MIGRATIONS_LOCK_TIMEOUT_FORCE`, which when set to True will guarantee that we always set the lock timeout for all operations. In cases where we would have previously set both a lock and statement timeout, the result is unchanged.

We specifically don't want to set a statement timeout on all operations. Some operations can take a long time (adding an index) and so we don't want to force this in all cases.

This pr vendors a bunch of code from `dzdm`. Ideally we could attempt to submit it upstream, but I think we'd need to tidy things up a little to do so.

### CODE REVIEW

File: src/sentry/db/postgres/schema.py [hunk 1]
```diff
@@ -97,6 +112,71 @@ def remove_field(self, model, field):
...
+        yield
+        if statement_timeout is not None:
+            PostgresDatabaseSchemaEditor.execute(
+                self,
+                self.sql_set_statement_timeout % {"statement_timeout": previous_statement_timeout},
+            )
+        if lock_timeout is not None:
+            PostgresDatabaseSchemaEditor.execute(
+                self, self.sql_set_lock_timeout % {"lock_timeout": previous_lock_timeout}
+            )
```
Comment (wedamija) [inline src/sentry/db/postgres/schema.py:178]: Copied from https://github.com/tbicr/django-pg-zero-downtime-migrations/blob/a96fb4dea8361857ca24b05e84a9ab6f612a3bb6/django_zero_downtime_migrations/backends/postgres/schema.py#L284-L335 and modified

**Comment by mwarkentin** [comment]
> This pr vendors a bunch of code from dzdm. Ideally we could attempt to submit it upstream, but I think we'd need to tidy things up a little to do so.

Worth opening an issue and describing our fix, I think. 👍🏻

**Review by vgrozdanic — APPROVED** [review]

**Review by markstory — APPROVED** [review]
Makes sense to me. The new force setting covers all operations that wouldn't have used timeouts otherwise.

**Review by armenzg — COMMENTED** [review]
lgtm

**Pull request merged**